### DISCUSSION TITLE
detecting responses/signals from external button box

### DISCUSSION THREAD

**Discussion by phileasheim** [jspsych/jsPsych#1748] [Q&A]
Hi everyone, 

we are currently working on moving our online-study back to the lab and we would love to continue conducting our experiments with jsPsych. Now we are struggling with detecting responses given with a response button via an input-output-device (button box) which is connected to the PC via USB. The PC recognizes the device as "data acquisition device". 
In the past we used the software presentation (NeuroBehavioral Systems), which allows you to define the button-box as input-channel and so to recognize the button-press as response. 
We've hoped, that we could somehow detect signals from the button box with jsPsych.ALL_KEYS, but made the experience that this is probably limited to the input from keyboard devices. 

Has anyone experiences with linking jsPsych-trials (like html-keyboard-response) to signals/responses from this kind of devices?

**Comment by jodeleeuw** [comment]
I've been able to connect to some external hardware to jsPsych (e.g., EEG), but it has usually required adding one more piece of software in the loop that has lower-level access to the hardware.

I don't know much about it, but it looks like there is an experimental api for USB devices: https://web.dev/usb/. Since you are in the lab and can control what browser you are using this option might be viable.

**Reply by phileasheim** [reply]
Thank you very much for your answer. We'll definitely check out the experimental api!

As you have mentioned connecting EEG to jsPsych: We've have actually also tried to send triggers to the EGG with the hardware-plugin by Daniel Rivas. We've installed his C++ program and also the chrome-extension but weren't able to connect to the parallel port. We know the port is working (we use an ANT Neuro EEG system) because we are able to send triggers with the presentation program I mentioned above. 

Do you have any further advice for getting this connection to work? We're also not sure about the right hex-number for the parallel port, although we've used the instructions of Daniel Rivas. As we interpret the extension, the connection never seemed successful. 

Additionally in our demo-code we've tried to use an on_start function to send the triggers, but we are also not sure if this is correct. 

    var stimuli = {
        type: 'image-keyboard-response',
        stimulus: 'stimuli/vis_right1.png',
        choices: jsPsych.ALL_KEYS,
        response_ends_trial: false,
        trial_duration: 1000,
        on_start: function () {
            document.dispatchEvent(new CustomEvent('jspsych', {
                detail:{
                    target:'parallel',
                    action:'trigger',
                    payload: 64         
                }
            }));
        }
    };
